loadtest/control/simulcontroller/controller.go

```diff
@@ -118,136 +122,191 @@ func (c *SimulController) Run() {
 			run:       c.searchChannels,
 			frequency: 0.5,
 		},
 		{
+			name:      "AddReaction",
 			run:       c.addReaction,
 			frequency: 0.5,
 		},
 		{
+			name:      "FullReload",
 			run:       c.fullReload,
 			frequency: 0.2,
 		},
 		{
+			name:      "CreateDirectChannel",
 			run:       c.createDirectChannel,
 			frequency: 0.25,
 		},
 		{
+			name:      "LogoutLogin",
 			run:       c.logoutLogin,
 			frequency: 0.1,
 		},
 		{
+			name:      "SearchUsers",
 			run:       searchUsers,
 			frequency: 0.1,
 		},
 		{
+			name:      "SearchPosts",
 			run:       searchPosts,
 			frequency: 0.1,
 		},
 		{
+			name:      "CreatePostReminder",
 			run:       c.createPostReminder,
 			frequency: 0.002,
 		},
 		{
+			name:      "EditPost",
 			run:       editPost,
 			frequency: 0.1,
 		},
 		{
+			name:      "DeletePost",
 			run:       deletePost,
 			frequency: 0.06,
 		},
 		{
+			name:      "UpdateCustomStatus",
 			run:       c.updateCustomStatus,
 			frequency: 0.05,
 		},
 		{
+			name:      "RemoveCustomStatus",
 			run:       c.removeCustomStatus,
 			frequency: 0.05,
 		},
 		{
+			name:      "CreateSidebarCategory",
 			run:       c.createSidebarCategory,
 			frequency: 0.06,
 		},
 		{
+			name:      "UpdateSidebarCategory",
 			run:       c.updateSidebarCategory,
 			frequency: 0.06,
 		},
 		{
+			name:      "SearchGroupChannels",
 			run:       searchGroupChannels,
 			frequency: 0.1,
 		},
 		{
+			name:      "CreateGroupChannel",
 			run:       c.createGroupChannel,
 			frequency: 0.05,
 		},
 		{
+			name:      "CreatePrivateChannel",
 			run:       createPrivateChannel,
 			frequency: 0.022,
 		},
 		{
+			name:      "CreatePublicChannel",
 			run:       control.CreatePublicChannel,
 			frequency: 0.011,
 		},
 		{
+			name:      "ViewGlobalThreads",
 			run:       c.viewGlobalThreads,
 			frequency: 5.4,
 		},
 		{
+			name:      "FollowThread",
 			run:       c.followThread,
 			frequency: 0.041,
 		},
 		{
+			name:      "UnfollowThread",
 			run:       c.unfollowThread,
 			frequency: 0.055,
 		},
 		{
+			name:      "ViewThread",
 			run:       c.viewThread,
 			frequency: 4.8,
 		},
 		{
+			name:      "MarkAllThreadsInTeamAsRead",
 			run:       c.markAllThreadsInTeamAsRead,
 			frequency: 0.013,
 		},
 		{
+			name:      "UpdateThreadRead",
 			run:       c.updateThreadRead,
 			frequency: 1.17,
 		},
 		{
+			name:      "GetInsights",
 			run:       c.getInsights,
 			frequency: 0.011,
 		},
 	}
+	for _, ua := range actions {
+		c.actionMap[ua.name] = ua
+	}
+
+	var injectedAction *userAction
 
 	for {
+		// run injected actions (if any) first
+		if injectedAction != nil {
+			if resp := injectedAction.run(c.user); resp.Err != nil {
+				c.status <- c.newErrorStatus(resp.Err)
+			} else {
+				c.status <- c.newInfoStatus(resp.Info)
+			}
+			injectedAction = nil
+		}
+
+		// check for injected actions
+		select {
+		case ia := <-c.injectedActionChan:
+			injectedAction = &ia
+			continue
+		default:
+		}
+
```

Comment:
This whole `for` is getting quite confusing. The structure is something like this, right?
```go
for {
  // 1. run any injected action
  // 2. check for new injected actions and skip the next iteration
  // 3. run a normal action
  // 4. wait for the idle time between actions OR receive a new injected action immediately and go to 1
}
```
I think 1+2 (in particular the `continue` at 2) and the OR in 4 are confusing. Now, I don't have good solutions:

I thought of moving 1+2 to a `checkAndRunInjectedAction` function, maybe swapping the order there: if injectedAction is nil, then read the channel and update injectedAction, then run it if not nil. But it's also quite confusing.

Any thoughts?

Reply:
I wonder if we could run the injected action instead of the picked one to simplify and have a single `action` variable. The rest of the logic doesn't really care whether it's injected or picked.

Reply:
I refactored the loop for SimulController and SimpleController.  I think this should address the concerns.